### PR TITLE
optimize reorder inside tensor_iterator

### PR DESCRIPTION
**Pull request by usstq**
### Details:
 - extract common logic for creating cached reorder primitive 
 - optimize edge mem copy logic inside tensor iterator with this cached reorder primitive
 - for RNNT , this logic can speed-up TI since it doesn't have to create reorder many times
 - removed a lot of superfluous shared_ptr wrappers on onednn handles

### Tickets:
 - CVS-85825

### CODE REVIEW

File: src/plugins/intel_cpu/src/nodes/common/reorder_prim.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+namespace ov {
+namespace intel_cpu {
+
+std::shared_ptr<dnnl::primitive> getReorderPrim(MultiCachePtr cache,
+                                                const dnnl::engine& engine,
+                                                const dnnl::memory::desc& src,
+                                                const dnnl::memory::desc& dest,
+                                                impl_desc_type* p_impl_type = nullptr);
+
+inline std::shared_ptr<dnnl::primitive> getReorderPrim(MultiCachePtr cache,
```
Comment (ceciliapeng2011) [inline src/plugins/intel_cpu/src/nodes/common/reorder_prim.h:21]: why do we need overload getReorderPrim instead of keeping just one?

**Review by ceciliapeng2011 — APPROVED** [review]
Verified it works good for ticket 87788 as well.

**Comment by dmitry-gorokhov** [comment]
@a-sidorova  @maxnick  could you please review?

File: src/plugins/intel_cpu/src/nodes/common/reorder_prim.h (hunk 1 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/common/reorder_prim.h:21]: Seems like we don't have always access to `engine` to set it as parameter

**Review by a-sidorova — APPROVED** [review]

File: src/plugins/intel_cpu/src/nodes/common/reorder_prim.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <ie_common.h>
+#include <node.h>
+
+#include <memory>
+
+namespace ov {
+namespace intel_cpu {
+
+std::shared_ptr<dnnl::primitive> getReorderPrim(MultiCachePtr cache,
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/common/reorder_prim.h:15]: Good idea, will remove superfluous shared_ptr

File: src/plugins/intel_cpu/src/nodes/common/reorder_prim.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/common/reorder_prim.h:21]: OK, will remove this overload

Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/common/reorder_prim.h:21]: Will remove extra overloads to simplify APIs

**Comment by usstq** [comment]
@maxnick @a-sidorova @ceciliapeng2011 inspired by @maxnick 's suggestion, I removed a lot of superfluous shared_ptr wrappers on onednn handle. it makes the code more clean and I believe also reduces a lot small mallocs at runtime, please review again, Thx!

File: src/plugins/intel_cpu/src/nodes/transpose.cpp [hunk 3]
```diff
@@ -164,29 +140,15 @@ void Transpose::prepareParams() {
...
-            impl_type = parse_impl_name(info);
-            return std::make_shared<dnnl::reorder>(pd);
-        };
-
-        auto cache = getRuntimeCache();
-        auto result = cache->getOrCreate(key, builder);
-
-        if (!result.first) {
+        auto result = getReorderPrim(getRuntimeCache(), getEngine(), src_blocked->GetPrimitive().get_desc(), dst_blocked->GetPrimitive().get_desc());
+        if (!result) {
```
Comment (maxnick) [inline src/plugins/intel_cpu/src/nodes/transpose.cpp:144]: Making `src_blocked` and `dst_blocked` memory objects is not necessary at all. Memory descriptors could be retrieved from the memory objects on edges directly. Actually this is the same thing that is fixed in PR https://github.com/openvinotoolkit/openvino/pull/14366. Could you please refactor this part as well?

Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/transpose.cpp]: OK, done!

Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/transpose.cpp]: @maxnick Please review again, Thx!

**Review by maxnick — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@usstq  Please update the PR to resolve conflicts

**Comment by usstq** [comment]
@dmitry-gorokhov Done.

**Comment by wenjiew** [comment]
Should we merge this PR?

**Review by ceciliapeng2011 — APPROVED** [review]

**Pull request merged**